### ISSUE TITLE
[QUESTION]More experts (more than 8, each EP group > 1), continuing pre-training of MoE with grouped_gemm results in abnormal initial loss (too high), around 6.x.

### ISSUE THREAD

**Issue by Emperorizzis** [NVIDIA/Megatron-LM#909]
Hi! Thank you for your open-source work!

When continuing the pre-training of [Qwen1.5-MoE-A2.7B](https://huggingface.co/Qwen/Qwen1.5-MoE-A2.7B/blob/main/config.json#L17) with the same configuration, using `SequentialMLP`, the **loss remains around 1.x**, but using `GroupedMLP`, the **loss is around 6.x**.

I noticed the structural differences between `GroupedMLP` and `SequentialMLP`. After converting the model weights from Huggingface to a format accepted by Megatron (`SequentialMLP`, and verified with a loss around 1.x), I used the following method to convert the Experts' weights:
```python
...
def convert_seq_mlp_to_grouped_gemm(ep_state_dict: dict, num_layers: int, num_local_experts: int):
    for l in tqdm(range(num_layers), desc='Transform experts format'):
        fc1_params = [f"decoder.layers.{l}.mlp.experts.local_experts.{e}.linear_fc1.weight" for e in range(num_local_experts)]
        w1 = torch.cat([ep_state_dict[p] for p in fc1_params], dim=0).transpose(0,1)
        for p in fc1_params:
            del ep_state_dict[p]
        ep_state_dict[f"decoder.layers.{l}.mlp.experts.weight1"] = w1
        fc2_params = [f"decoder.layers.{l}.mlp.experts.local_experts.{e}.linear_fc2.weight" for e in range(num_local_experts)]
        w2 = torch.cat([ep_state_dict[p] for p in fc2_params], dim=1).transpose(0,1)
        for p in fc2_params:
            del ep_state_dict[p]
        ep_state_dict[f"decoder.layers.{l}.mlp.experts.weight2"] = w2
...
```

The model weights load correctly, but the **initial loss has increased to 6.x**.

After comparing, I found that the output values from the forward computations of `SequentialMLP` experts and `GroupedMLP` experts are different, while everything else is the same. Could there be an issue with the model weight conversion logic, or is it something else wrong?

The version of `grouped_gemm` being used is the latest version ([main branch](https://github.com/fanshiqing/grouped_gemm)).

**Looking forward to your reply!**

**Comment by Ethan-yt** [comment]
Let's find the shape of each tensor for FC1:

- **FC1 Weight of SeqMLP**: (ffn_hidden, hidden)
- **After concatenation**: (expert_num * ffn_hidden, hidden)
- **After transposition**: (hidden, expert_num * ffn_hidden)

This is correct according to the definition of GroupedMLP. However, if you examine the `forward` method of GroupedMLP, you'll find:

```python
w1 = self.weight1.view(self.num_local_experts, self.config.hidden_size, -1)
w2 = self.weight2.view(self.num_local_experts, -1, self.config.hidden_size)
```

This code actually accepts a tensor with the shape (expert_num, hidden, ffn_hidden). Therefore, you should first create a tensor with this shape and then reshape it to (hidden, expert_num * ffn_hidden). 

The discrepancy between the definition and usage is somewhat confusing. Recently, a new module called `TEGroupedMLP` was introduced. Switching to this module might be a better solution.

---

Does this meet your needs?

**Comment by Emperorizzis** [comment]
> Let's find the shape of each tensor for FC1:
> 
> * **FC1 Weight of SeqMLP**: (ffn_hidden, hidden)
> * **After concatenation**: (expert_num * ffn_hidden, hidden)
> * **After transposition**: (hidden, expert_num * ffn_hidden)
> 
> This is correct according to the definition of GroupedMLP. However, if you examine the `forward` method of GroupedMLP, you'll find:
> 
> ```python
> w1 = self.weight1.view(self.num_local_experts, self.config.hidden_size, -1)
> w2 = self.weight2.view(self.num_local_experts, -1, self.config.hidden_size)
> ```
> 
> This code actually accepts a tensor with the shape (expert_num, hidden, ffn_hidden). Therefore, you should first create a tensor with this shape and then reshape it to (hidden, expert_num * ffn_hidden).
> 
> The discrepancy between the definition and usage is somewhat confusing. Recently, a new module called `TEGroupedMLP` was introduced. Switching to this module might be a better solution.
> 
> Does this meet your needs?

Thank you very much for your answer, the problem has been resolved now! 🤗

**Comment by derange-alembic** [comment]
> Let's find the shape of each tensor for FC1:
> 
> * **FC1 Weight of SeqMLP**: (ffn_hidden, hidden)
> * **After concatenation**: (expert_num * ffn_hidden, hidden)
> * **After transposition**: (hidden, expert_num * ffn_hidden)
> 
> This is correct according to the definition of GroupedMLP. However, if you examine the `forward` method of GroupedMLP, you'll find:
> 
> w1 = self.weight1.view(self.num_local_experts, self.config.hidden_size, -1)
> w2 = self.weight2.view(self.num_local_experts, -1, self.config.hidden_size)
> This code actually accepts a tensor with the shape (expert_num, hidden, ffn_hidden). Therefore, you should first create a tensor with this shape and then reshape it to (hidden, expert_num * ffn_hidden).
> 
> The discrepancy between the definition and usage is somewhat confusing. Recently, a new module called `TEGroupedMLP` was introduced. Switching to this module might be a better solution.
> 
> Does this meet your needs?

Hi @Ethan-yt , can I ask why the weights in GroupedMLP need to be defined and used in different shapes? Can I simply define the weight as (expert_num, hidden, ffn_hidden) ? Besides, if I want to manipulate the weight of a specific expert, what is the correct way to do that ? Thank you !

**Comment by Ethan-yt** [comment]
> > Let's find the shape of each tensor for FC1:
> > 
> > * **FC1 Weight of SeqMLP**: (ffn_hidden, hidden)
> > * **After concatenation**: (expert_num * ffn_hidden, hidden)
> > * **After transposition**: (hidden, expert_num * ffn_hidden)
> > 
> > This is correct according to the definition of GroupedMLP. However, if you examine the `forward` method of GroupedMLP, you'll find:
> > w1 = self.weight1.view(self.num_local_experts, self.config.hidden_size, -1)
> > w2 = self.weight2.view(self.num_local_experts, -1, self.config.hidden_size)
> > This code actually accepts a tensor with the shape (expert_num, hidden, ffn_hidden). Therefore, you should first create a tensor with this shape and then reshape it to (hidden, expert_num * ffn_hidden).
> > The discrepancy between the definition and usage is somewhat confusing. Recently, a new module called `TEGroupedMLP` was introduced. Switching to this module might be a better solution.
> > Does this meet your needs?
> 
> Hi [@Ethan-yt](https://github.com/Ethan-yt) , can I ask why the weights in GroupedMLP need to be defined and used in different shapes? Can I simply define the weight as (expert_num, hidden, ffn_hidden) ? Besides, if I want to manipulate the weight of a specific expert, what is the correct way to do that ? Thank you !

`GroupedMLP ` is designed to be used in that way. I think you can try `TEGroupedMLP`.
